src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -387,332 +368,165 @@ public bool TryGetValue(string key, out object value)
-        {
-            public abstract int Count { get; }
-
-            public abstract KeyValuePair<string, object> this[int index] { get; set; }
-
-            public abstract void Upgrade(ref Storage storage);
-
-            public abstract bool TryGetValue(string key, out object value);
-
-            public abstract bool ContainsKey(string key);
-
-            public abstract bool TrySetValue(string key, object value);
+            value = default;
+            return false;
         }
 
-        internal class ListStorage : Storage
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsureCapacity(int capacity)
         {
-            private KeyValuePair<string, object>[] _items;
-            private int _count;
-
-            private static readonly KeyValuePair<string, object>[] _emptyArray = new KeyValuePair<string, object>[0];
-
-            public ListStorage()
+            if (_propertyStorage != null || _arrayStorage.Length < capacity)
             {
-                _items = _emptyArray;
+                EnsureCapacitySlow(capacity);
             }
+        }
 
-            public ListStorage(int capacity)
+        private void EnsureCapacitySlow(int capacity)
+        {
+            if (_propertyStorage != null)
             {
-                if (capacity == 0)
-                {
-                    _items = _emptyArray;
-                }
-                else
-                {
-                    _items = new KeyValuePair<string, object>[capacity];
-                }
-            }
+                var storage = _propertyStorage;
+                capacity = Math.Max(storage.Properties.Length, capacity);
+                var array = new KeyValuePair<string, object>[capacity];
 
-            public ListStorage(ListStorage other)
-            {
-                if (other.Count == 0)
+                for (var i = 0; i < storage.Properties.Length; i++)
                 {
-                    _items = _emptyArray;
-                }
-                else
-                {
-                    _items = new KeyValuePair<string, object>[other.Count];
-                    for (var i = 0; i < other.Count; i++)
-                    {
-                        this.Add(other[i]);
-                    }
+                    var property = storage.Properties[i];
+                    array[i] = new KeyValuePair<string, object>(property.Name, property.GetValue(storage.Value));
                 }
+                
+                _arrayStorage = array;
+                _propertyStorage = null;
+                return;
             }
 
-            public int Capacity => _items.Length;
-
-            public override int Count => _count;
-
-            public override KeyValuePair<string, object> this[int index]
+            if (_arrayStorage.Length < capacity)
             {
-                get
+                capacity = _arrayStorage.Length == 0 ? 4 : _arrayStorage.Length * 2;
```

Comment:
4?

Reply:
It's a good number yeah? THis should be the constant value I defined elsewhere. 

A route with areas will typically need 4 entries, most routes need fewer.